File: 2022/day2.js
const { getData } = require('./data');

async function part1() {
    const data = await getData(2);
    const dict = {
        'A': { 'X': 4, 'Y': 8, 'Z': 3 },
        'B': { 'X': 1, 'Y': 5, 'Z': 9 },
        'C': { 'X': 7, 'Y': 2, 'Z': 6 },
    };
    return data.split('\n').reduce((a, c) => a + dict[c.split(' ')[0]][c.split(' ')[1]], 0);
}

async function part2() {
    const data = await getData(2);
    const dict = {
        'A': { 'X': 3, 'Y': 4, 'Z': 8 },
        'B': { 'X': 1, 'Y': 5, 'Z': 9 },
        'C': { 'X': 2, 'Y': 6, 'Z': 7 },
    };
    return data.split('\n').reduce((a, c) => a + dict[c.split(' ')[0]][c.split(' ')[1]], 0);
}

part1().then(res => console.log('Part 1\n', res));

part2().then(res => console.log('Part 2\n', res));

File: 2023/day2.py
from data import get_data

data = get_data(2)

def is_valid_game(game):
    """Returns the ID if the game was valid, or 0 if not (so the sum is unaffected)"""
    game_id, draws_str = game.split(': ')
    id = game_id.split(' ')[1]
    draws = draws_str.split('; ')
    for draw in draws:
        colours = draw.split(', ')
        for draw_colour in colours:
            number, colour = draw_colour.split(' ')
            if colour == 'red' and int(number) > 12:
                return 0
            if colour == 'green' and int(number) > 13:
                return 0
            if colour == 'blue' and int(number) > 14:
                return 0
    return int(id)

def part1():
    ids = [is_valid_game(game) for game in data.split('\n')]
    return sum(ids)

print(part1())

def get_game_power(game):
    red = 0
    green = 0
    blue = 0
    _id, draws_str = game.split(': ')
    draws = draws_str.split('; ')
    for draw in draws:
        colours = draw.split(', ')
        for draw_colour in colours:
            number, colour = draw_colour.split(' ')
            if colour == 'red' and int(number) > red:
                red = int(number)
            if colour == 'green' and int(number) > green:
                green = int(number)
            if colour == 'blue' and int(number) > blue:
                blue = int(number)
    return red * green * blue

def part2():
    powers = [get_game_power(game) for game in data.split('\n')]
    return sum(powers)

print(part2())
